src/main/java/org/zalando/fahrschein/NakadiClientBuilder.java

```diff
@@ -0,0 +1,62 @@
+package org.zalando.fahrschein;
+
+import org.springframework.http.client.ClientHttpRequestFactory;
+import org.springframework.http.client.SimpleClientHttpRequestFactory;
+
+import java.net.URI;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.util.Optional.ofNullable;
+
+public final class NakadiClientBuilder {
+    public static final int DEFAULT_CONNECT_TIMEOUT = 500;
+    public static final int DEFAULT_READ_TIMEOUT = 60 * 1000;
+
+    private final URI baseUri;
+    private AccessTokenProvider accessTokenProvider;
+    private ClientHttpRequestFactory clientHttpRequestFactory;
+    private CursorManager cursorManager;
+
+    NakadiClientBuilder(final URI baseUri) {
+        this.baseUri = checkNotNull(baseUri, "Base URI should not be null");
+    }
+
+    private NakadiClientBuilder(URI baseUri, AccessTokenProvider accessTokenProvider, ClientHttpRequestFactory clientHttpRequestFactory, CursorManager cursorManager) {
+        this.baseUri = baseUri;
+        this.accessTokenProvider = accessTokenProvider;
+        this.clientHttpRequestFactory = clientHttpRequestFactory;
+        this.cursorManager = cursorManager;
+    }
+
+    public NakadiClientBuilder withAccessTokenProvider(AccessTokenProvider accessTokenProvider) {
+        return new NakadiClientBuilder(baseUri, accessTokenProvider, clientHttpRequestFactory, cursorManager);
```

Comment:
Should the builder be stateful or not? Looks like a mixture here. I'd expect a builder to be stateful, i.e. 
```
this.accessTokenProvider = accessTokenProvider
return this
```